feat(ducks-schematics): add effects + selectors option

skydever opened the pull request:
Hi there! 

This adds the option `effects` to the `duck` schematic (default: `true`).

When turned on, it will create a `<duck name>.effects.ts` file including the basic setup of the effects class.
I also added a prompt for it + I added a prompt for the `barrel` option.

Can you review @GregOnNet?

ps: I want to add the possibility for selectors too, but within a separate pr.
**edit**: added the selectors option here too

skydever committed:
feat(ducks-schematics): add effects option

skydever commented:
about the selctors, they depend on the state setup - going this way with the schematics?

```typescript
const visitCounter = createFeatureSelector<CounterState>('counter');
```

skydever commented:
I fixed the imports and only create the effect at the duck class if the flag is set...

skydever committed:
feat(ducks-schematics): add selectors option

GregOnNet commented on ngrx/ducks-schematics/src/collection/duck/files/__name@dasherize@if-flat__/__name@dasherize__.effects.ts.template:
```diff
@@ -0,0 +1,20 @@
+
+@Injectable()
+export class <%= classify(name) %>Effects {
+  @Effect()
```
Can you add a comment that this is an example?
I often see that generated code is not replaced by developers and remains unless it is already released.

```ts
/**
 * Generated example
 *
 * You may want to replace this Effect with your own implementation.
 */

```

skydever replied:
I added the comment, and I will try to add the Effect to the nearest module and/or add an extra parameter for it - a behavior like `ng generate module` would be great I think ...

skydever replied:
kinda tricky to check and add it to an existing `EffectsModule.forFeature([SomeOtherEffects])` entry in the nearest, or via params passed, module - else import `EffectsModule` from ngrx and add the whole entry `EffectsModule.forFeature([MyDuckEffects])`, also have to check for `EffectsModule.forRoot(...)` I guess (add as root if no root was found?) ... is there a public API to help out here? 🤔 I am copying + adopting the angular schematics sources, feels a bit strange ... but I think I can handle it ... maybe it is a good idea to also check the ngrx schematics sources ... definitely a good place to learn a lot I guess ...

... and how to handle the selectors? leave it like it is, or register the reducer too, with some options? maybe splitting the schematic is a better option?

GregOnNet replied:
## Effect

It should be the same behaviour as adding a generated Component to a module file.

I did not implement this before but the API you may have to look for is called `recorder` (see: https://www.softwarearchitekt.at/post/2018/03/20/custom-schematics-part-iv-frictionless-library-setup-with-the-angular-cli-and-schematics.aspx).

We can also postpone this feature to a later release. :-)

What do you think?

## Selectors

For now, I would leave it as is.
In general, I prefer to have separate schematics for the effect and the selector.

skydever replied:
thx for the link :+1: I will check it out and come back afterwards (soon hopefully)

skydever replied:
... seems like borrowing from the angular schematics sources is kind of *best practice* until there is a public api, does not feel that strange anymore 😄

## Effect

If `effects` is set (there is a prompt for it as well), it will add `EffectsModule.forFeature([DuckEffects])` to the nearest `NgModule` (like the default component, pipe... schematics do) including the required typescript imports (`import { EffectsModule } from '@ngrx/effects';` + the relative import for the `DuckEffects`).

Optionally you can pass a `module` to explicitly define the declaring `NgModule` for the effect.

If you want to skip the declaration at a `NgModule`, you can pass `skipImport`.

Now it should behave like adding a generated component.

The schematic does not check if there is already a import statement for an effect (eg: `EffectsModule.forFeature([OtherEffects])`), it will always add a new line `EffectsModule.forFeature([DuckEffects])` instead of adding it to an existing `EffectsModule.forFeature([OtherEffects, DuckEffects])` statement. Maybe I can handle this, but I guess this is something for a later release :-)

### Selectors

I did not touch the selectors.

Having separate schematics for selectors + effects would be nice and I think I can work on it, but I guess this is also something for a later release :-) Will look into this in the near future.

GregOnNet reviewed (changes requested):
Hey, thanks for your PR.

I do not see the code where the generated Effect is added to the corresponding module.
Can you have look how this could be done?

skydever committed:
feat(ducks-schematics): add comment to gen. effect

skydever committed:
feat(duck-schematics): add module + skipImport ...

... for the NgModule import of the effect.
auto-detect nearest module for the effect,
if no module is provided.
skipImport: no NgModule import at all.

skydever reviewed (commented):
added the `NgModule` import

GregOnNet commented:
Hi @skydever,

thank you so much for the update.
I noticed one small bug.

If I create a Duck in a module that already contains an EffectModule, a second EffectsModule is created instead of adding the generated Effect to the existing one.

Can you have a look at this?

![image](https://user-images.githubusercontent.com/444278/56360322-97e02e00-61e4-11e9-85da-c24c80f3805a.png)

skydever committed:
style(prettier): format code

skydever committed:
fix(ducks-schematics): add to EffectsModule ...

... if the declaring module already contains an EffectsModule

skydever commented:
Hi @GregOnNet 

You're welcome, I am happy to contribute and learn new things while doing it 👍

I added the check to add the effect to an existing EffectsModule, if there is one, else it I will add the EffectsModule itself. Have to do some refactoring, but it should work 🎉

GregOnNet commented:
Hi,

I have tested the schematic.
Everything looks fine.

I already released the new version!

Thank you so much.

skydever commented:
awesome :+1:
